internal/cache/file/downloader/job.go

```diff
@@ -226,196 +225,186 @@ func (job *Job) notifySubscribers() {
+			return err
+		}
+
+		start += maxRead
+		if start == newReaderLimit {
+			err = newReader.Close()
+			if err != nil {
+				logger.Errorf("Job:%p (%s:/%s) error while closing reader: %v", job, job.bucket.Name(), job.object.Name, err)
+			}
+			newReader = nil
+		}
+
+		err = job.updateStatusOffset(start)
+		if err != nil {
+			return err
+		}
+	}
+	return
+}
+
+// cleanUpDownloadAsyncJob is a helper function which performs clean up tasks
+// for the async job and this should be called at the end of async job.
+//
+// Acquires and releases LOCK(job.mu)
+func (job *Job) cleanUpDownloadAsyncJob() {
+	// Close the job.doneCh, clear the cancelFunc & cancelCtx and call the
+	// remove job callback function.
+	job.cancelFunc()
+	close(job.doneCh)
+
+	job.mu.Lock()
+	if job.removeJobCallback != nil {
+		job.removeJobCallback()
+		job.removeJobCallback = nil
+	}
+	job.cancelCtx, job.cancelFunc = nil, nil
+	job.mu.Unlock()
+}
+
 // downloadObjectAsync downloads the backing GCS object into a file as part of
 // file cache using NewReader method of gcs.Bucket.
 //
 // Note: There can only be one async download running for a job at a time.
 // Acquires and releases LOCK(job.mu)
 func (job *Job) downloadObjectAsync() {
-	// Close the job.doneCh, clear the cancelFunc & cancelCtx and call the
-	// remove job callback function in any case - completion/failure.
-	defer func() {
-		job.cancelFunc()
-		close(job.doneCh)
-
-		job.mu.Lock()
-		if job.removeJobCallback != nil {
-			job.removeJobCallback()
-			job.removeJobCallback = nil
-		}
-		job.cancelCtx, job.cancelFunc = nil, nil
-		job.mu.Unlock()
-	}()
+	// Cleanup the async job in all cases - completion/failure/invalidation.
+	defer job.cleanUpDownloadAsyncJob()
 
 	// Create, open and truncate cache file for writing object into it.
 	cacheFile, err := cacheutil.CreateFile(job.fileSpec, os.O_TRUNC|os.O_WRONLY)
 	if err != nil {
 		err = fmt.Errorf("downloadObjectAsync: error in creating cache file: %w", err)
-		job.failWhileDownloading(err)
+		job.updateStatusAndNotifySubscribers(Failed, err)
 		return
 	}
 	defer func() {
 		err = cacheFile.Close()
 		if err != nil {
 			err = fmt.Errorf("downloadObjectAsync: error while closing cache file: %w", err)
-			job.failWhileDownloading(err)
+			job.updateStatusAndNotifySubscribers(Failed, err)
 		}
 	}()
 
-	notifyInvalid := func() {
-		job.mu.Lock()
-		job.status.Name = Invalid
-		job.notifySubscribers()
-		job.mu.Unlock()
+	// Both parallel and non-parallel download functions supports cancellation in
+	// case of job's cancellation.
+	if job.fileCacheConfig.EnableParallelDownloads {
+		err = job.parallelDownloadObjectToFile(cacheFile)
+	} else {
+		err = job.downloadObjectToFile(cacheFile)
 	}
 
-	var newReader io.ReadCloser
-	var start, end, sequentialReadSize, newReaderLimit int64
-	end = int64(job.object.Size)
-	sequentialReadSize = int64(job.sequentialReadSizeMb) * cacheutil.MiB
-
-	for {
-		select {
-		case <-job.cancelCtx.Done():
```

Comment:
where are we checking if context is cancelled?

Reply:
It's not required because both parallelDownloadObjectToFile & downloadObjectToFile returns err (context.Canceled type) if context is canceled. This is because the `newReader` that we get as output of NewReader call is derived from job.cancelCtx and hence `_, err = io.CopyN(offsetWriter, newReader, maxRead)` fails in that case with context canceled error. This is equivalent to what was there before.